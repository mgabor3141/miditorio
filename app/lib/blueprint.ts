import { arrayChunks, encodeBlueprint } from '@/app/lib/utils'
import signals from '@/app/lib/data/signals-dlc.json'
import { FinalInstruments } from '@/app/lib/song-to-blueprint'

const qualities = signals
  .filter(({ type }) => type === 'quality')
  .map(({ name }) => name)

// This is a reserved signal because of the playback circuit
const RESERVED = ['signal-green']

const signalsWithQuality = signals
  .flatMap((signal) =>
    qualities.map((quality) => ({
      ...signal,
      quality,
    })),
  )
  .filter(
    ({ name, quality }) => !(quality === 'normal' && RESERVED.includes(name)),
  )

export const toBlueprint = ({
  tickCombinatorValues,
  dataCombinatorValues,
  instruments,
}: {
  tickCombinatorValues: number[]
  dataCombinatorValues: number[]
  instruments: FinalInstruments
}) => {
  console.log(
    `Got ${tickCombinatorValues.length} signals. ` +
      `(${signalsWithQuality.length} total signals available.)`,
  )

  // TODO
  tickCombinatorValues = tickCombinatorValues.slice(
    0,
    signalsWithQuality.length,
  )
  dataCombinatorValues = dataCombinatorValues.slice(
    0,
    signalsWithQuality.length,
  )

  const entities: {
    name: string
    entity_number: number
    position: { x: number; y: number }
  }[] = []

  /**
   * source_entity_number
   * source_wire_connector_id
   * target_entity_number
   * target_wire_connector_id
   */
  const wires: number[][] = []

  // Add speakers and their combinators
  // WARNING! There is code that relies on speakers and their combinators being
  //  added first to the entity array!
  Object.values(instruments)
    .map(({ instrument, volume }, instrumentNumber) => [
      {
        name: 'programmable-speaker',
        position: {
          x: 0,
          y: -2.5,
        },
        control_behavior: {
          circuit_condition: {
            first_signal: {
              type: 'item',
              name: 'programmable-speaker',
            },
            constant: 0,
            comparator: '>',
          },
          circuit_parameters: {
            signal_value_is_pitch: true,
            instrument_id: Number(instrument.id),
            note_id: 0,
          },
        },
        parameters: {
          playback_volume: instrument.volumeCorrection * volume,
          playback_mode: 'surface',
          allow_polyphony: true,
        },
      },
      {
        name: 'decider-combinator',
        position: {
          x: 0,
          y: -1,
        },
        control_behavior: {
          decider_conditions: {
            conditions: [
              {
                first_signal: {
                  type: 'virtual',
                  name: 'signal-each',
                },
                // Instrument address
                constant:
                  (instrumentNumber + 1) <<
                  (6 + 8 * ((instrumentNumber + 1) % 2)),
                comparator: '=',
                first_signal_networks: {
                  red: true,
                  green: false,
                },
              },
            ],
            outputs: [
              {
                signal: {
                  name: 'programmable-speaker',
                },
                networks: {
                  red: false,
                  green: true,
                },
              },
            ],
          },
        },
      },
    ])
    .forEach(([speakerEntity, combinatorEntity], instrumentNumber) => {
      ;[speakerEntity, combinatorEntity].forEach((entity) =>
        entities.push({
          ...entity,
          entity_number: entities.length,
          position: {
            x: entity.position.x + Math.floor(entities.length / 4),
            y: entity.position.y + Math.floor((entities.length % 4) / 2) * 3,
          },
        }),
      )

      wires.push([
        entities.length - 1, // combinator entity number
        4, // wire_connector_id
        entities.length - 2, // speaker entity number
        2, // wire_connector_id
      ])

      if (instrumentNumber >= 2) {
        wires.push([
          entities.length - 1, // combinator entity number
          1, // wire_connector_id
          entities.length - 5, // combinator before previous entity number
          1, // wire_connector_id
        ])
        wires.push([
          entities.length - 1, // combinator entity number
          2, // wire_connector_id
          entities.length - 5, // previous combinator entity number
          2, // wire_connector_id
        ])
      }
    })

  // Add static logic
  ;[
    {
      // entity_number: 1,
      name: 'decider-combinator',
      position: {
        x: -3,
        y: 0.5,
      },
      direction: 4,
      control_behavior: {
        decider_conditions: {
          conditions: [
            {
              first_signal: {
                type: 'virtual',
                name: 'signal-green',
              },
              comparator: '>',
              first_signal_networks: {
                red: false,
                green: true,
              },
            },
          ],
          outputs: [
            {
              signal: {
                type: 'virtual',
                name: 'signal-everything',
              },
            },
          ],
        },
      },
      player_description:
        'Clock\n\n[virtual-signal=signal-green] is the time (number of ticks) since the start of the song.',
    },
    {
      // entity_number: 2,
      name: 'decider-combinator',
      position: {
        x: -3,
        y: -0.5,
      },
      direction: 4,
      control_behavior: {
        decider_conditions: {
          conditions: [
            {
              first_signal: {
                type: 'virtual',
                name: 'signal-each',
              },
              second_signal: {
                type: 'virtual',
                name: 'signal-green',
              },
              comparator: '=',
              first_signal_networks: {
                red: true,
                green: false,
              },
            },
          ],
          outputs: [
            {
              signal: {
                type: 'virtual',
                name: 'signal-each',
              },
              networks: {
                red: false,
                green: true,
              },
            },
          ],
        },
      },
      player_description:
        'Let each signal through on the [color=green]green wire[/color] where the value of that same signal on the [color=red]red wire[/color] is equal to the [virtual-signal=signal-green] (time) signal.\n\nThis makes [virtual-signal=signal-green] a reserved signal and must not be present in the memory.\n\nOutput is any number of signals, where each signal contains one or two events, bit packed. Each event is made up of an instrument address and a note value.',
    },
    {
      // entity_number: 3,
      name: 'constant-combinator',
      position: {
        x: -3,
        y: 1.5,
      },
      direction: 8,
      control_behavior: {
        sections: {
          sections: [
            {
              index: 1,
              filters: [
                {
                  index: 1,
                  type: 'virtual',
                  name: 'signal-green',
                  quality: 'normal',
                  comparator: '=',
                  count: 1,
                },
              ],
            },
          ],
        },
        is_on: false,
      },
      player_description:
        '[virtual-signal=signal-green]\n[virtual-signal=signal-green][virtual-signal=signal-green]\n[virtual-signal=signal-green][virtual-signal=signal-green][virtual-signal=signal-green]\n[virtual-signal=signal-green][virtual-signal=signal-green]\n[virtual-signal=signal-green]\n\nCreated using miditorio.com v2',
    },
    {
      // entity_number: 5, -> 4
      name: 'arithmetic-combinator',
      position: {
        x: -1,
        y: -1.5,
      },
      direction: 4,
      control_behavior: {
        arithmetic_conditions: {
          first_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
          second_constant: 4194304,
          operation: '/',
          output_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
        },
      },
      player_description:
        'Get note value for event 2 from each signal\n\nResult: 6 bits\n0000 1111 1100 0000 0000 0000 0000 0000\n\nOperation: Integer divide by\n0000 0000 0100 0000 0000 0000 0000 0000',
    },

    {
      // entity_number: 6, -> 5
      name: 'arithmetic-combinator',
      position: {
        x: -1,
        y: -0.5,
      },
      direction: 4,
      control_behavior: {
        arithmetic_conditions: {
          first_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
          second_constant: 4177920,
          operation: 'AND',
          output_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
        },
      },
      player_description:
        'Get instrument address for event 2 from each signal\n\nResult: 8 bits\n0000 0000 0011 1111 1100 0000 0000 0000\n\nResult is not shifted to zero, instruments check against the unshifted number',
    },
    {
      // entity_number: 7, -> 6
      name: 'arithmetic-combinator',
      position: {
        x: -1,
        y: 0.5,
      },
      direction: 4,
      control_behavior: {
        arithmetic_conditions: {
          first_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
          second_constant: 16320,
          operation: 'AND',
          output_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
        },
      },
      player_description:
        'Get instrument address for event 1 from each signal\n\nResult: 8 bits\n0000 0000 0000 0000 0011 1111 1100 0000\n\nResult is not shifted to zero, instruments check against the unshifted number',
    },
    {
      // entity_number: 8, -> 7
      name: 'arithmetic-combinator',
      position: {
        x: -1,
        y: 1.5,
      },
      direction: 4,
      control_behavior: {
        arithmetic_conditions: {
          first_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
          second_constant: 64,
          operation: '%',
          output_signal: {
            type: 'virtual',
            name: 'signal-each',
          },
        },
      },
      player_description:
        'Get note value for event 1 from each signal\n\nResult: 6 bits\n0000 0000 0000 0000 0000 0000 0011 1111\n\nOperation: Modulo\n0000 0000 0000 0000 0000 0000 0100 0000',
    },
  ].forEach((entity) => {
    entities.push({ ...entity, entity_number: entities.length })
  })

  {
    // Helper function to convert from the original entity numbers to
    //  the new ones in the `entities` array
    const e = (originalNumber: number): number =>
      entities.length - 8 + originalNumber
    wires.push(
      [e(4), 2, e(5), 2],
      [e(1), 1, e(1), 3],
      [e(1), 2, e(3), 2],
      [e(1), 4, e(2), 2],
      [e(2), 4, e(5), 2],
      [e(5), 2, e(6), 2],
      [e(6), 2, e(7), 2],
      [e(4), 4, 1, 2], // To first speaker green
      [e(5), 3, 1, 1], // To first speaker red
      [e(6), 3, 3, 1], // To second speaker red
      [e(7), 4, 3, 2], // To second speaker green
    )
  }
  const clockCombinatorEntityNumber = entities.length - 6

  // Add data combinators
  ;[tickCombinatorValues, dataCombinatorValues]
    .map((values, isDataCombinator) => ({
      name: 'constant-combinator',
      direction: 4,
      control_behavior: {
        sections: {
          sections: arrayChunks(values, 1000).map((section, sectionIndex) => ({
            index: sectionIndex + 1,
            filters: section.map((value, valueIndex) => ({
              index: valueIndex + 1,
              ...signalsWithQuality[sectionIndex * 1000 + valueIndex],
              count: value,
            })),
          })),
        },
      },
      player_description: isDataCombinator
        ? 'Instrument and note data\n\n' +
          'Each signal value represents one or two song events. Each signal has a corresponding signal in the other combinator whose value contains the timing information for those events.'
        : 'Timing data\n\n' +
          'Each signal value is a point in time where one or more events happen. Each signal has a corresponding signal in the other combinator whose value contains the information for those events.',
    }))
    .forEach((entity, combinatorNumber) => {
      entities.push({
        ...entity,
        entity_number: entities.length,
        position: { x: -5, y: combinatorNumber - 1 },
      })
    })

  wires.push(
    [entities.length - 2, 1, clockCombinatorEntityNumber, 1],
    [entities.length - 1, 2, clockCombinatorEntityNumber, 2],
  )

  const blueprint = {
    blueprint: {
      icons: [
        {
          signal: {
            name: 'programmable-speaker',
          },
          index: 1,
        },
      ],
      entities,
      wires,
      item: 'blueprint',
      version: 281483568218115,
    },
  }

  return encodeBlueprint(blueprint)
}
