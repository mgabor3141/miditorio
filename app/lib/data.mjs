// Factorio instruments

export function FactorioInstrument(
  name,
  id,
  base,
  range_octaves,
  default_volume,
) {
  this.name = name
  this.id = id
  this.base = base
  this.range = range_octaves * 12

  if (default_volume !== undefined) this.default_volume = default_volume
  else this.default_volume = 1
}

FactorioInstrument.prototype.convert = function (note) {
  return note - this.base
}

FactorioInstrument.prototype.checkRange = function (note, getDirection) {
  const pitch = note.convert()

  if (getDirection === false || getDirection === undefined)
    return pitch > 0 && pitch <= this.range

  if (pitch <= 0) return { direction: 'below', delta: -pitch + 1 }
  if (pitch > this.range)
    return { direction: 'above', delta: pitch - this.range }
  return true
}

export const factorio_instrument = {
  Piano: new FactorioInstrument('Piano', 3, 40, 4),
  Bass: new FactorioInstrument('Bass', 4, 28, 3),
  Lead: new FactorioInstrument('Lead', 5, 28, 3),
  Saw: new FactorioInstrument('Saw', 6, 28, 3, 0.8),
  Square: new FactorioInstrument('Square', 7, 28, 3, 0.5),
  Celesta: new FactorioInstrument('Celesta', 8, 64, 3),
  Vibraphone: new FactorioInstrument('Vibraphone', 9, 52, 3),
  Plucked: new FactorioInstrument('Plucked', 10, 52, 3),
  SteelDrum: new FactorioInstrument('SteelDrum', 11, 40, 3),
  ReverseCymbal: new FactorioInstrument('ReverseCymbal', 2),
  DrumKit: new FactorioInstrument('DrumKit', 2),
  Exclude: new FactorioInstrument('Exclude', -1),
}

factorio_instrument['ReverseCymbal'].convert = function (_note) {
  return 13
}
factorio_instrument['ReverseCymbal'].checkRange = function (
  _note,
  _getDirection,
) {
  return true
}

factorio_instrument['DrumKit'].convert = function (note) {
  if (drumkit_map[note] === undefined) {
    // console.log('GM DrumKit ' + note + ' not found.')
    return 0
  } else {
    return drumkit_map[note]
  }
}
factorio_instrument['DrumKit'].checkRange = function (_note, _getDirection) {
  return true
}

factorio_instrument['Exclude'].convert = function (_note) {
  return 0
}
factorio_instrument['Exclude'].checkRange = function (_note, _getDirection) {
  return true
}

export function getFactorioInstrument(programCode) {
  programCode++
  let factorioInstrumentName = 'Exclude'

  if (programCode === 0) {
    // Drumkit
    factorioInstrumentName = 'DrumKit'
  } else if (programCode <= 8) {
    // Piano
    factorioInstrumentName = 'Piano'
  } else if (programCode <= 16) {
    // Chromatic Percussion
    if (programCode === 9) factorioInstrumentName = 'Celesta'
    else factorioInstrumentName = 'Vibraphone'
  } else if (programCode <= 24) {
    // Organ
    factorioInstrumentName = 'Square'
  } else if (programCode <= 32) {
    // Guitar
    factorioInstrumentName = 'Saw'
  } else if (programCode <= 40) {
    // Bass
    factorioInstrumentName = 'Bass'
  } else if (programCode <= 48) {
    // Strings
    factorioInstrumentName = 'Lead'
  } else if (programCode <= 56) {
    // Ensemble
    factorioInstrumentName = 'Lead'
  } else if (programCode <= 64) {
    // Brass
    factorioInstrumentName = 'Piano'
  } else if (programCode <= 72) {
    // Reed
    factorioInstrumentName = 'Piano'
  } else if (programCode <= 80) {
    // Pipe
    factorioInstrumentName = 'Piano'
  } else if (programCode <= 88) {
    // Synth Lead
    if (programCode === 81) factorioInstrumentName = 'Square'
    else if (programCode === 82) factorioInstrumentName = 'Saw'
    else factorioInstrumentName = 'Lead'
  } else if (programCode <= 96) {
    // Synth Pad
    factorioInstrumentName = 'Vibraphone'
  } else if (programCode <= 104) {
    // Synth Effects
    factorioInstrumentName = 'Saw'
  } else if (programCode <= 112) {
    // Ethnic
    factorioInstrumentName = 'Piano'
  } else if (programCode <= 120) {
    // Percussive
    if (programCode === 120) factorioInstrumentName = 'ReverseCymbal'
    else factorioInstrumentName = 'SteelDrum'
  } else if (programCode <= 128) {
    // Sound Effects
    factorioInstrumentName = 'Exclude'
  }

  return factorio_instrument[factorioInstrumentName]
}

const drumkit_map = {
  27: 9, // High Q -> high q
  35: 2, // Acoustic Bass Drum -> kick 2
  36: 1, // Bass Drum 1 -> kick 1
  37: 10, // Side Stick -> perc 1
  38: 5, // Accoustic Snare -> snare 2
  39: 14, // Hand Clap -> clap
  40: 4, // Electric Snare -> snare 1
  41: 4, // Low Floor Tom -> snare 1
  42: 6, // Closed Hi Hat -> hat 1
  43: 4, // High Floor Tom -> snare 1
  44: 7, // Pedal Hi Hat -> hat 2
  45: 4, // Low Tom -> snare 1
  46: 7, // Open Hi Hat -> hat 2
  47: 4, // Low-Mid Tom -> snare 1
  48: 4, // Hi-Mid Tom -> snare 1
  49: 12, // Crash Cymbal 1 -> crash
  50: 4, // High Tom -> snare 1
  51: 6, // Ride Cymbal -> hat 1
  52: 12, // Chinese Cymbal -> crash
  53: 16, // Ride Bell -> cowbell
  54: 15, // Tambourine -> shaker
  55: 12, // Splash Cymbal -> crash
  56: 16, // Cowbell -> cowbell
  57: 12, // Crash Cymbal 2 -> crash
  59: 6, // Ride Cymbal 2 -> hat 1
  69: 15, // Cabasa -> shaker
  75: 10, // Claves -> perc 2
  76: 10, // Hi Wood Block -> perc 2
  77: 10, // Low Wood Block -> perc 2
  81: 17, // Open Triangle -> triangle
}

// GM instrument names
const midi_instrument = [
  'Acoustic Grand Piano',
  'Bright Acoustic Piano',
  'Electric Grand Piano',
  'Honky-tonk Piano',
  'Electric Piano 1',
  'Electric Piano 2',
  'Harpsichord',
  'Clavi',
  'Celesta',
  'Glockenspiel',
  'Music Box',
  'Vibraphone',
  'Marimba',
  'Xylophone',
  'Tubular Bells',
  'Dulcimer',
  'Drawbar Organ',
  'Percussive Organ',
  'Rock Organ',
  'Church Organ',
  'Reed Organ',
  'Accordion',
  'Harmonica',
  'Tango Accordion',
  'Acoustic Guitar (nylon)',
  'Acoustic Guitar (steel)',
  'Electric Guitar (jazz)',
  'Electric Guitar (clean)',
  'Electric Guitar (muted)',
  'Overdriven Guitar',
  'Distortion Guitar',
  'Guitar harmonics',
  'Acoustic Bass',
  'Electric Bass (finger)',
  'Electric Bass (pick)',
  'Fretless Bass',
  'Slap Bass 1',
  'Slap Bass 2',
  'Synth Bass 1',
  'Synth Bass 2',
  'Violin',
  'Viola',
  'Cello',
  'Contrabass',
  'Tremolo Strings',
  'Pizzicato Strings',
  'Orchestral Harp',
  'Timpani',
  'String Ensemble 1',
  'String Ensemble 2',
  'SynthStrings 1',
  'SynthStrings 2',
  'Choir Aahs',
  'Voice Oohs',
  'Synth Voice',
  'Orchestra Hit',
  'Trumpet',
  'Trombone',
  'Tuba',
  'Muted Trumpet',
  'French Horn',
  'Brass Section',
  'SynthBrass 1',
  'SynthBrass 2',
  'Soprano Sax',
  'Alto Sax',
  'Tenor Sax',
  'Baritone Sax',
  'Oboe',
  'English Horn',
  'Bassoon',
  'Clarinet',
  'Piccolo',
  'Flute',
  'Recorder',
  'Pan Flute',
  'Blown Bottle',
  'Shakuhachi',
  'Whistle',
  'Ocarina',
  'Lead 1 (square)',
  'Lead 2 (sawtooth)',
  'Lead 3 (calliope)',
  'Lead 4 (chiff)',
  'Lead 5 (charang)',
  'Lead 6 (voice)',
  'Lead 7 (fifths)',
  'Lead 8 (bass + lead)',
  'Pad 1 (new age)',
  'Pad 2 (warm)',
  'Pad 3 (polysynth)',
  'Pad 4 (choir)',
  'Pad 5 (bowed)',
  'Pad 6 (metallic)',
  'Pad 7 (halo)',
  'Pad 8 (sweep)',
  'FX 1 (rain)',
  'FX 2 (soundtrack)',
  'FX 3 (crystal)',
  'FX 4 (atmosphere)',
  'FX 5 (brightness)',
  'FX 6 (goblins)',
  'FX 7 (echoes)',
  'FX 8 (sci-fi)',
  'Sitar',
  'Banjo',
  'Shamisen',
  'Koto',
  'Kalimba',
  'Bag pipe',
  'Fiddle',
  'Shanai',
  'Tinkle Bell',
  'Agogo',
  'Steel Drums',
  'Woodblock',
  'Taiko Drum',
  'Melodic Tom',
  'Synth Drum',
  'Reverse Cymbal',
  'Guitar Fret Noise',
  'Breath Noise',
  'Seashore',
  'Bird Tweet',
  'Telephone Ring',
  'Helicopter',
  'Applause',
  'Gunshot',
]
midi_instrument[-1] = 'Drum Kit'

// GM percussion names
const _midi_percussion = {
  35: 'Acoustic Bass Drum',
  36: 'Bass Drum 1',
  37: 'Side Stick',
  38: 'Acoustic Snare',
  39: 'Hand Clap',
  40: 'Electric Snare',
  41: 'Low Floor Tom',
  42: 'Closed Hi Hat',
  43: 'High Floor Tom',
  44: 'Pedal Hi-Hat',
  45: 'Low Tom',
  46: 'Open Hi-Hat',
  47: 'Low-Mid Tom',
  48: 'Hi-Mid Tom',
  49: 'Crash Cymbal 1',
  50: 'High Tom',
  51: 'Ride Cymbal 1',
  52: 'Chinese Cymbal',
  53: 'Ride Bell',
  54: 'Tambourine',
  55: 'Splash Cymbal',
  56: 'Cowbell',
  57: 'Crash Cymbal 2',
  58: 'Vibraslap',
  59: 'Ride Cymbal 2',
  60: 'Hi Bongo',
  61: 'Low Bongo',
  62: 'Mute Hi Conga',
  63: 'Open Hi Conga',
  64: 'Low Conga',
  65: 'High Timbale',
  66: 'Low Timbale',
  67: 'High Agogo',
  68: 'Low Agogo',
  69: 'Cabasa',
  70: 'Maracas',
  71: 'Short Whistle',
  72: 'Long Whistle',
  73: 'Short Guiro',
  74: 'Long Guiro',
  75: 'Claves',
  76: 'Hi Wood Block',
  77: 'Low Wood Block',
  78: 'Mute Cuica',
  79: 'Open Cuica',
  80: 'Mute Triangle',
  81: 'Open Triangle',
}
