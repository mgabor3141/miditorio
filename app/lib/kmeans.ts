export type ClusterResult = {
  clusters: number[][]
  centers: number[]
}

// Main function to automatically determine the number of clusters
export function autoCluster(
  data: number[],
  threshold = 0.05,
  maxK = 10,
): ClusterResult {
  const maxClusters = Math.min(maxK, data.length)
  let prevMeanDeviation = Infinity
  let bestClusters: ClusterResult | null = null

  for (let k = 1; k <= maxClusters; k++) {
    const clustersResult = kMeansClustering(data, k)
    const meanDeviation = computeMeanDeviation(clustersResult)

    if (prevMeanDeviation !== Infinity) {
      const meanDeviationDecrease = prevMeanDeviation - meanDeviation

      if (
        prevMeanDeviation === 0 ||
        meanDeviationDecrease / prevMeanDeviation < threshold
      ) {
        // Stop iterating if the decrease in mean deviation is less than the threshold
        break
      }
    }

    prevMeanDeviation = meanDeviation
    bestClusters = clustersResult
  }

  if (bestClusters === null) {
    // Fallback in case clustering fails
    return kMeansClustering(data, 1)
  }

  return bestClusters
}

// K-means clustering function with k-means++ initialization
export function kMeansClustering(data: number[], k: number): ClusterResult {
  // Initialize centers using k-means++ initialization
  let centers = initializeCenters(data, k)

  let clusters: number[][] = []
  const maxIterations = 100
  let iteration = 0
  let centersChanged = true

  while (centersChanged && iteration < maxIterations) {
    // Assign points to the nearest centers
    clusters = assignPointsToClusters(data, centers)

    // Update centers based on current clusters
    const newCenters = updateCenters(clusters, data)

    // Check if centers have changed significantly
    centersChanged = centersHaveChanged(centers, newCenters)

    centers = newCenters
    iteration++
  }

  return { clusters, centers }
}

// Initialize cluster centers using k-means++ algorithm
function initializeCenters(data: number[], k: number): number[] {
  const centers: number[] = []

  // Step 1: Choose the first center randomly from the data points
  const firstCenter = data[Math.floor(Math.random() * data.length)]
  centers.push(firstCenter)

  // Steps 2-4: Choose the remaining centers
  while (centers.length < k) {
    // Compute squared distances to the nearest center
    const distances = data.map((point) => {
      const minDistance = Math.min(
        ...centers.map((center) => Math.abs(point - center)),
      )
      return minDistance ** 2
    })

    const sumDistances = distances.reduce((sum, d) => sum + d, 0)

    if (sumDistances === 0) {
      // All points are identical; randomly select remaining centers
      while (centers.length < k) {
        const randomCenter = data[Math.floor(Math.random() * data.length)]
        centers.push(randomCenter)
      }
      break
    }

    // Compute probabilities for selecting the next center
    const probabilities = distances.map((d) => d / sumDistances)

    // Compute cumulative probabilities
    const cumulativeProbabilities =
      computeCumulativeProbabilities(probabilities)

    // Select a new center based on the cumulative probabilities
    const rand = Math.random()
    let newCenterIndex = cumulativeProbabilities.findIndex((p) => rand < p)
    if (newCenterIndex === -1) {
      newCenterIndex = cumulativeProbabilities.length - 1
    }
    centers.push(data[newCenterIndex])
  }

  return centers
}

// Assign data points to the nearest cluster centers
function assignPointsToClusters(data: number[], centers: number[]): number[][] {
  const clusters = centers.map(() => [] as number[])
  for (const point of data) {
    let minDistance = Infinity
    let closestCenterIndex = -1
    for (let i = 0; i < centers.length; i++) {
      const distance = Math.abs(point - centers[i])
      if (distance < minDistance) {
        minDistance = distance
        closestCenterIndex = i
      }
    }
    clusters[closestCenterIndex].push(point)
  }
  return clusters
}

// Update cluster centers based on the mean of assigned points
function updateCenters(clusters: number[][], data: number[]): number[] {
  return clusters.map((cluster) => {
    if (cluster.length > 0) {
      // Calculate the new center as the mean of the cluster points
      return cluster.reduce((a, b) => a + b, 0) / cluster.length
    } else {
      // If a cluster is empty, reinitialize its center randomly
      return data[Math.floor(Math.random() * data.length)]
    }
  })
}

// Check if cluster centers have changed significantly
function centersHaveChanged(
  oldCenters: number[],
  newCenters: number[],
  tolerance = 1e-6,
): boolean {
  for (let i = 0; i < oldCenters.length; i++) {
    if (Math.abs(oldCenters[i] - newCenters[i]) > tolerance) {
      return true
    }
  }
  return false
}

// Compute cumulative probabilities for selecting centers
function computeCumulativeProbabilities(probabilities: number[]): number[] {
  const cumulativeProbabilities = []
  let sum = 0
  for (const p of probabilities) {
    sum += p
    cumulativeProbabilities.push(sum)
  }
  return cumulativeProbabilities
}

// Compute the mean deviation of all points from their cluster centers
function computeMeanDeviation(clusterResult: ClusterResult): number {
  let totalDeviation = 0
  let totalPoints = 0

  for (let i = 0; i < clusterResult.clusters.length; i++) {
    const cluster = clusterResult.clusters[i]
    const center = clusterResult.centers[i]
    for (const point of cluster) {
      totalDeviation += Math.abs(point - center)
      totalPoints += 1
    }
  }

  return totalPoints > 0 ? totalDeviation / totalPoints : 0
}

export function roundToNearestClusterCenter(
  value: number,
  clusterCenters: number[],
): { closestCenter: number; closestCenterNumber: number } {
  // Find the cluster center that is closest to the input value
  clusterCenters.sort()
  let closestCenterNumber = 0
  let minDistance = Math.abs(value - clusterCenters[0])

  for (let i = 1; i < clusterCenters.length; i++) {
    const center = clusterCenters[i]
    const distance = Math.abs(value - center)
    if (distance < minDistance) {
      minDistance = distance
      closestCenterNumber = i
    }
  }

  return {
    closestCenter: clusterCenters[closestCenterNumber],
    closestCenterNumber,
  }
}
