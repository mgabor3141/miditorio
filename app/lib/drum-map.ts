import { FactorioDrumSound } from '@/app/lib/data/factorio-drumkit-sounds-by-id'
import { MidiNote } from 'tone/build/esm/core/type/NoteUnits'
import { gmPercussionToFactorioDrumkit } from '@/app/lib/data/gm-percussion-to-factorio-drumkit'
import { noteToGmPercussion } from '@/app/lib/data/gm-percussion-note-names'

export type DrumMap = (note: MidiNote) => FactorioDrumSound | undefined

export type DrumMapOverrides = Record<string, FactorioDrumSound>

export const drumMapWithOverrides =
  (drumMap: DrumMap, overrides?: DrumMapOverrides): DrumMap =>
  (note) => {
    if (overrides && note in overrides) return overrides[note]

    return drumMap(note)
  }

export const defaultDrumMap: DrumMap = (note: MidiNote) =>
  gmPercussionToFactorioDrumkit[
    noteToGmPercussion[note.toString() as keyof typeof noteToGmPercussion]
  ]
