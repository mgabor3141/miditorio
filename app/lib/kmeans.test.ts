import { describe, expect, test } from 'vitest'
import { autoCluster } from '@/app/lib/kmeans'
import { readFile, readdir } from 'fs/promises'
import { Midi } from '@tonejs/midi'
import groupBy from 'lodash.groupby'

describe.sequential('K-means', async () => {
  const testFiles = (await readdir('test-data/')).filter(
    // Exclude files that are trivial to cluster
    (file) => !['bach.mid', 'sea.mid'].includes(file),
  )

  test('random mock', () => {
    expect(Math.random()).toMatchInlineSnapshot(`0.03475257847458124`)
  })

  test.for(testFiles)(
    'auto clustering %s',
    { timeout: 60_000 },
    async (testFile, { expect }) => {
      const numberOfTestsToAverage = 100

      const file = await readFile(`test-data/${testFile}`)
      const song = new Midi(file)

      const tracks = song.tracks.filter(
        (track) => track.notes.length && !track.instrument.percussion,
      )

      expect(
        tracks.map(({ name, notes }) => {
          const clusters = new Array(numberOfTestsToAverage)
            .fill(undefined)
            .map(
              () =>
                autoCluster({
                  data: notes.map(({ velocity }) => velocity),
                }).clusters.length,
            )

          return {
            _trackName: name.trim(),
            _trackNotes: notes.length,
            min: clusters.reduce((a, b) => Math.min(a, b)),
            max: clusters.reduce((a, b) => Math.max(a, b)),
            average:
              clusters.reduce((acc, v) => acc + v, 0) / numberOfTestsToAverage,
            median: clusters
              .toSorted()
              .at(Math.floor(numberOfTestsToAverage / 2)),
            mode: Object.entries(groupBy(clusters)).toSorted(
              ([_1, v1], [_2, v2]) => v2.length - v1.length,
            )[0][0],
          }
        }),
      ).toMatchSnapshot()
    },
  )
})
