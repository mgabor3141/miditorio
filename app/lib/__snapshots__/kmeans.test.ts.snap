// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`K-means > auto clustering hall.mid 1`] = `
[
  {
    "_trackName": "Piano",
    "_trackNotes": 724,
    "average": 7.43,
    "max": 9,
    "median": 7,
    "min": 5,
    "mode": "8",
  },
  {
    "_trackName": "",
    "_trackNotes": 727,
    "average": 8.82,
    "max": 10,
    "median": 9,
    "min": 6,
    "mode": "9",
  },
]
`;

exports[`K-means > auto clustering lune.mid 1`] = `
[
  {
    "_trackName": "Piano",
    "_trackNotes": 1498,
    "average": 5.58,
    "max": 7,
    "median": 6,
    "min": 4,
    "mode": "6",
  },
]
`;

exports[`K-means > auto clustering toto.mid 1`] = `
[
  {
    "_trackName": "",
    "_trackNotes": 714,
    "average": 5.32,
    "max": 7,
    "median": 5,
    "min": 3,
    "mode": "5",
  },
  {
    "_trackName": "",
    "_trackNotes": 483,
    "average": 4.34,
    "max": 5,
    "median": 4,
    "min": 3,
    "mode": "4",
  },
  {
    "_trackName": "",
    "_trackNotes": 66,
    "average": 3.03,
    "max": 4,
    "median": 3,
    "min": 3,
    "mode": "3",
  },
  {
    "_trackName": "",
    "_trackNotes": 608,
    "average": 2.24,
    "max": 3,
    "median": 2,
    "min": 2,
    "mode": "2",
  },
  {
    "_trackName": "",
    "_trackNotes": 508,
    "average": 6.02,
    "max": 9,
    "median": 6,
    "min": 4,
    "mode": "6",
  },
  {
    "_trackName": "",
    "_trackNotes": 136,
    "average": 3.96,
    "max": 4,
    "median": 4,
    "min": 3,
    "mode": "4",
  },
  {
    "_trackName": "",
    "_trackNotes": 116,
    "average": 1,
    "max": 1,
    "median": 1,
    "min": 1,
    "mode": "1",
  },
  {
    "_trackName": "",
    "_trackNotes": 81,
    "average": 4,
    "max": 4,
    "median": 4,
    "min": 4,
    "mode": "4",
  },
  {
    "_trackName": "",
    "_trackNotes": 85,
    "average": 4.05,
    "max": 5,
    "median": 4,
    "min": 4,
    "mode": "4",
  },
]
`;
